ocaml/xapi/storage_smapiv1_wrapper.ml

```diff
@@ -529,7 +541,8 @@ functor
                     )
                 with e ->
                   if not allow_leak then (
-                    ignore (Vdi.add_leaked dp vdi_t) ;
+                    Sr.add_or_replace vdi (Vdi.add_leaked dp vdi_t) sr_t ;
```

Comment:
Just to help myself to understand...
This is only for SMAPIv1.
There are two cases in which the `destroy_datapath_nolock` will be called:
1. `DP.destroy_sr`: this happens when `SR.detach` or `VBD.unplug`.
2. `remove_datapaths_andthen_nolock`: this happens before all VDI related operations.
In the case 1, it is to deliberately remove the datapath. Now with the fix, if the removal failed due to exceptions from e.g. SM, the datapath will be recorded as leaked in both memory and file. So that in case 2, it could be identified and be removed again. Furthermore, the removal failure again in case 2 will fail the `then` VDI operation and eventually expose the error. I think this is expected as it would be bad to ignore the error which might cause more issues.

Reply:
But in `VM.pool_migrate` or `VDI.pool_migrate` case, the toolstack restart only remove the DPs used in migrations, would it remove the leaked DPs of the VM? The VM doesn't resident on the source host actually. But the leaked DPs still blocks the VM attach on the destination host.

Reply:
>This is only for SMAPIv1.

On master (XS 8), yes, as the Storage_smapiv1_wrapper covers only SMAPIv1 SRs. Before, the module was called Storage_impl and covered both SMAPIv1 and SMAPIv3. The change was made following an update on how qemu-dp is used inside SMAPIv3 plugins.

Reply:
>But in VM.pool_migrate or VDI.pool_migrate case, the toolstack restart only remove the DPs used in migrations, would it remove the leaked DPs of the VM? The VM doesn't resident on the source host actually. But the leaked DPs still blocks the VM attach on the destination host.

If a dp was leaked during a VM.reboot, then it will now be properly recorded as such. A following VM.pool_migrate includes a VDI.deactivate call on the source host, where now the wrapper will notice the leaked dp and will get rid of it before proceeding. This happens before VDI.activate is called on the remote host.